// @rebass/mdx/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _tag = require('@mdx-js/tag');

var _markdown = require('@rebass/markdown');

var _markdown2 = _interopRequireDefault(_markdown);

var _rebass = require('rebass');

var _LiveEditor = require('./LiveEditor');

var _LiveEditor2 = _interopRequireDefault(_LiveEditor);

var _LivePreview = require('./LivePreview');

var _LivePreview2 = _interopRequireDefault(_LivePreview);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var pre = function pre(props) {
  return props.children;
};
var code = function code(_ref) {
  var children = _ref.children,
      _ref$className = _ref.className,
      className = _ref$className === undefined ? '' : _ref$className,
      scope = _ref.scope,
      props = _objectWithoutProperties(_ref, ['children', 'className', 'scope']);

  var lang = className.replace(/^language\-/, '');
  var type = lang.charAt(0);
  var code = _react2.default.Children.toArray(children).join('\n');

  switch (type) {
    case '.':
      return _react2.default.createElement(_LiveEditor2.default, { code: code, scope: scope });
    case '!':
      return _react2.default.createElement(_LivePreview2.default, { code: code, scope: scope });
    default:
      return _react2.default.createElement(_rebass.Pre, {
        p: 3,
        mt: 4,
        mb: 4,
        bg: 'gray',
        children: children
      });
  }
};

var RebassMDXProvider = function (_React$Component) {
  _inherits(RebassMDXProvider, _React$Component);

  function RebassMDXProvider() {
    var _ref2;

    var _temp, _this, _ret;

    _classCallCheck(this, RebassMDXProvider);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref2 = RebassMDXProvider.__proto__ || Object.getPrototypeOf(RebassMDXProvider)).call.apply(_ref2, [this].concat(args))), _this), _this.state = {
      components: {}
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(RebassMDXProvider, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          theme = _props.theme,
          children = _props.children;
      var components = this.state.components;


      return _react2.default.createElement(
        _rebass.Provider,
        { theme: theme },
        _react2.default.createElement(
          _tag.MDXProvider,
          { components: components },
          this.props.children
        )
      );
    }
  }]);

  return RebassMDXProvider;
}(_react2.default.Component);

RebassMDXProvider.getDerivedStateFromProps = function (props, state) {
  var components = _extends({}, (0, _markdown2.default)(props.props), {
    code: code,
    pre: pre
  }, props.components);
  if (state.components === components) return null;
  return { components: components };
};

exports.default = RebassMDXProvider;
