// react-scope-provider/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withScope = exports.ScopeConsumer = exports.ScopeProvider = exports.Scope = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var Scope = exports.Scope = _react2.default.createContext(null);

var ScopeProvider = function ScopeProvider(_ref) {
  var scope = _ref.scope,
      props = _objectWithoutProperties(_ref, ['scope']);

  return _react2.default.createElement(Scope.Provider, _extends({}, props, {
    value: scope
  }));
};

exports.ScopeProvider = ScopeProvider;
ScopeProvider.propTypes = {
  scope: _propTypes2.default.object.isRequired
};

var ScopeConsumer = function ScopeConsumer(_ref2) {
  var _ref2$defaultScope = _ref2.defaultScope,
      defaultScope = _ref2$defaultScope === undefined ? {} : _ref2$defaultScope,
      props = _objectWithoutProperties(_ref2, ['defaultScope']);

  return _react2.default.createElement(
    Scope.Consumer,
    null,
    function (maybeScope) {
      return maybeScope ? props.children(maybeScope) : _react2.default.createElement(
        Scope.Provider,
        { value: defaultScope },
        _react2.default.createElement(Scope.Consumer, props)
      );
    }
  );
};

exports.ScopeConsumer = ScopeConsumer;
ScopeConsumer.propTypes = {
  defaultScope: _propTypes2.default.object
};

var withScope = function withScope(Component) {
  return function (_ref3) {
    var defaultScope = _ref3.defaultScope,
        props = _objectWithoutProperties(_ref3, ['defaultScope']);

    return _react2.default.createElement(
      ScopeConsumer,
      { defaultScope: defaultScope },
      function (scope) {
        return _react2.default.createElement(Component, _extends({}, props, {
          scope: scope
        }));
      }
    );
  };
};
exports.withScope = withScope;
